cmd/snap/debug: add "snap debug seeding" command for preseeding debugging (4/N)

anonymouse64 opened the pull request:
This is based on top of https://github.com/snapcore/snapd/pull/9035.

This adds `snap debug seeding`, which in the happy case where there was no system-key difference, will have output like this:

```
seeded:            true
preseeded:         true
image-preseeding:  9.318206712s
seed-completion:   3.872508077s
```

and in the case where there is a difference will look something like this:

```
seeded:            true
preseeded:         true
image-preseeding:  9.318206712s
seed-completion:   3.872508077s
preseed-system-key: {
  "apparmor-features": [
    "caps",
    "dbus",
    "domain",
    "file",
    "mount",
    "namespaces",
    "network",
    "network_v8",
    "policy",
    "ptrace",
    "query",
    "rlimit",
    "signal"
  ],
  "apparmor-parser-features": [
    "unsafe"
  ],
  "apparmor-parser-mtime": 1589907589,
  "build-id": "cb94e5eeee4cf7ecda53f8308a984cb155b55732",
  "cgroup-version": "1",
  "nfs-home": false,
  "overlay-root": "",
  "seccomp-compiler-version": "e6e309ad8aee052e5aa695dfaa040328ae1559c5 2.4.3 9b218ef9a4e508dd8a7f848095cb8875d10a4bf28428ad81fdc3f8dac89108f7 bpf-actlog",
  "seccomp-features": [
    "allow",
    "errno",
    "kill_process",
    "kill_thread",
    "log",
    "trace",
    "trap",
    "user_notif"
  ],
  "version": 10
}
seed-restart-system-key: {
  "apparmor-features": [
    "caps",
    "dbus",
    "domain",
    "file",
    "mount",
    "namespaces",
    "network",
    "policy",
    "ptrace",
    "query",
    "rlimit",
    "signal"
  ],
  "apparmor-parser-features": [
    "unsafe"
  ],
  "apparmor-parser-mtime": 1589907589,
  "build-id": "cb94e5eeee4cf7ecda53f8308a984cb155b55732",
  "cgroup-version": "1",
  "nfs-home": false,
  "overlay-root": "",
  "seccomp-compiler-version": "e6e309ad8aee052e5aa695dfaa040328ae1559c5 2.4.3 9b218ef9a4e508dd8a7f848095cb8875d10a4bf28428ad81fdc3f8dac89108f7 bpf-actlog",
  "seccomp-features": [
    "allow",
    "errno",
    "kill",
    "log",
    "trace",
    "trap",
    "user_notif"
  ],
  "version": 10
}
```

anonymouse64 commented:
GitHub Actions / spread is confused with the state of the PR again, closing and re-opening

bboozzoo commented on cmd/snap/cmd_debug_seeding_test.go:
```diff
@@ -0,0 +1,399 @@
+        "seed-restart-time": "2019-07-04T19:14:10.548793375-05:00",
+        "seed-start-time": "0001-01-01T00:00:00Z",
+        "seed-time": "2019-07-04T19:16:10.548793375-05:00",
+		"seeded": true,
+		"preseeded": true
```
```suggestion
	"seeded": true,
	"preseeded": true
```

bboozzoo commented on cmd/snap/cmd_debug_seeding.go:
```diff
@@ -0,0 +1,136 @@
+	var seedDuration time.Duration
+	if resp.Preseeded {
+		preseedDuration := resp.PreseedTime.Sub(resp.PreseedStartTime)
+		fmt.Fprintf(w, "image-preseeding:\t%v\n", preseedDuration)
```
Maybe we could `.Round(time.Millisecond)` here and when printing seed completion.

bboozzoo commented on cmd/snap/cmd_debug_seeding.go:
```diff
@@ -0,0 +1,136 @@
+		// exported) and return that back to us as an interface{} again, which
+		// we can then use to compare the two system keys, and then finally
+		// display the originally marshalled json
+		seedRestartSkJSON, err := json.MarshalIndent(resp.SeedRestartSystemKey, "", "  ")
```
Maybe you could define resp.SeedRestartSystemKey and resp.PreseedSystemKey as `*json.RawMessage`, then do:

```
seedSk, err := interfaces.UnmarshalJSONSystemKey(bytes.NewReader(*resp.SeedRestartSystemKey))
if err != nil {
	return err
}
preseedSk, err := interfaces.UnmarshalJSONSystemKey(bytes.NewReader(*resp.PreseedSystemKey))
if err != nil {
	return err
}

match, err := interfaces.SystemKeysMatch(preseedSk, seedSk)
if err != nil {
	return err
}
if !match {
	// MarshalIndent(pressedSk) and MarshalIndent(seedSk)
	// fprintf
}
```

bboozzoo commented on overlord/devicestate/devicemgr.go:
```diff
@@ -435,6 +435,26 @@ func (m *DeviceManager) ensureOperational() error {
+var startTime time.Time
+
+func init() {
+	startTime = time.Now()
```
I think this should be set in in DeviceManager when devicestate.Manager() is called.

anonymouse64 replied:
all the devicemgr bits here are actually from the other PR, https://github.com/snapcore/snapd/pull/9035, so I will address this feedback there

pedronis replied:
This was my original implementation. I actually measured and on my machine there's a 120ms between the init and Manager. So I think I would prefer to keep it this way, it's messier but gives a slightly better number.

pedronis commented on cmd/snap/cmd_debug_seeding.go:
```diff
@@ -0,0 +1,136 @@
+		SeedStartTime        time.Time   `json:"seed-start-time,omitempty"`
+		SeedRestartTime      time.Time   `json:"seed-restart-time,omitempty"`
+		SeedTime             time.Time   `json:"seed-time,omitempty"`
+		PreseedSystemKey     interface{} `json:"preseed-system-key,omitempty"`
```
these maybe should *json.RawMessage instead, then below can use json.Indent

pedronis commented on cmd/snap/cmd_debug_seeding.go:
```diff
@@ -0,0 +1,136 @@
+	} else {
+		seedDuration = resp.SeedTime.Sub(resp.SeedStartTime)
+	}
+	fmt.Fprintf(w, "seed-completion:\t%v\n", seedDuration)
```
we should maybe print:

seed-completion: -

instead if seeded is false?

pedronis commented on cmd/snap/cmd_debug_seeding.go:
```diff
@@ -0,0 +1,136 @@
+	// for the preseeded case, we use the seed-restart-time as the start time
+	// to show how long we spent only after booting the preseeded image
+	var seedDuration time.Duration
+	if resp.Preseeded {
```
we probably want to Truncate or Round the durations to ms for printing

pedronis commented on cmd/snap/cmd_debug_seeding.go:
```diff
@@ -0,0 +1,136 @@
+	if resp.Preseeded && resp.SeedRestartSystemKey != nil && resp.PreseedSystemKey != nil {
+		// only show them if they don't match
+
+		// we have to play a bit of a dance here, first marshalling the
```
see the other approach of using *json.RawMessage instead,  the end result is probably the same in terms of unmarshalings done but is maybe a bit cleaner

bboozzoo commented on overlord/devicestate/devicemgr.go:
```diff
@@ -435,6 +435,26 @@ func (m *DeviceManager) ensureOperational() error {
+	startTime = time.Now()
+}
+
+func (m *DeviceManager) setTimeOnce(name string, t time.Time) error {
```
Could this be a helper on state, eg. state.SetOnce("key", value) ?

anonymouse64 replied:
it certainly cloud be, @pedronis what do you think ( and also note that like the other comment this really applies to #9035 not this PR)

pedronis replied:
it seems an ok idea, but maybe follow up material,  it's also the first use case for this, so not sure it's worth adding to state yet.

bboozzoo commented on overlord/devicestate/devicemgr.go:
```diff
@@ -435,6 +435,26 @@ func (m *DeviceManager) ensureOperational() error {
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if !prev.IsZero() {
```
`if err == ErrNoState` should be enough, unless we're worried that a zero time might get stored in the state?

anonymouse64 replied:
I think that was the concern yes

pedronis replied:
if we move it to State we'll have to leave with the other semantics though. TBH I think I wrote this way just because, I probably copied some other code with that error handling and built on top.

bboozzoo commented on tests/lib/preseed.sh:
```diff
@@ -39,24 +39,54 @@ umount_ubuntu_image() {
+    local SEED_SNAPS_DIR="$IMAGE_MOUNTPOINT/var/lib/snapd/seed/snaps"
+
+    # need remarshal for going from json to yaml and back for seed manipulation
+    if ! command -v json2yaml || ! command -v yaml2json; then
```
This should be a trivial bit of Python and PyYAML should be already present in the host system, maybe we could add a tool for that?

anonymouse64 replied:
sure, if you want to refactor this go for it, I am already using remarshal and jq things like this elsewhere so this was quicker and easier for me to write, but I'm not opposed to something in Python as long as somebody else writes it :-)

anonymouse64 commented:
I addressed all the feedback here and made more cases show "-" when the time is missing or not calculatable, however for this to work IRL, the daemon will need to start outputting `*time.Time` too, however I'd rather do that in a separate PR since it's an orthogonal point (clients can decode into *time.Time while the server is outputting time.Time, and vice versa)

stolowski commented:
Thank you for this!
